server/protocols/jmap/src/main/java/org/apache/james/jmap/model/Keywords.java

```diff
@@ -0,0 +1,189 @@
+
+        public KeywordsFactory throwOnUnsupportedKeyword() {
+            validator = Optional.of(keywords -> Preconditions.checkArgument(
+                keywords.stream().allMatch(Keyword::isNonExposedImapKeyword), "Does not allow to update 'Deleted' or 'Recent' flag"));
+            return this;
+        }
+
+        public KeywordsFactory filterUnsupportedFlags() {
+            filter = Optional.of(keyword -> keyword.isNonExposedImapKeyword());
+            return this;
+        }
+
+        public Keywords fromSet(Set<Keyword> setKeywords) {
+            validator.orElse(keywords -> {})
+                .validate(setKeywords);
+
+            return new Keywords(setKeywords.stream()
+                .filter(filter.orElse(keyword -> true))
+                .collect(Guavate.toImmutableSet()));
+        }
+
+        @VisibleForTesting
+        Keywords fromMap(Map<String, Boolean> mapKeywords) {
+            Preconditions.checkArgument(mapKeywords.values()
+                .stream()
+                .noneMatch(keywordValue -> !keywordValue), "Keyword must be true");
+            Set<Keyword> setKeywords = mapKeywords.keySet()
+                .stream()
+                .map(Keyword::new)
+                .collect(Guavate.toImmutableSet());
+
+            return fromSet(setKeywords);
+        }
+
+        @VisibleForTesting
+        Keywords fromOldKeyword(OldKeyword oldKeyword) {
+            ImmutableSet.Builder<Keyword> builder = ImmutableSet.builder();
+            if (oldKeyword.isAnswered().isPresent() && oldKeyword.isAnswered().get()) {
+                builder.add(Keyword.ANSWERED);
+            }
+            if (oldKeyword.isDraft().isPresent() && oldKeyword.isDraft().get()) {
+                builder.add(Keyword.DRAFT);
+            }
+            if (oldKeyword.isFlagged().isPresent() && oldKeyword.isFlagged().get()) {
+                builder.add(Keyword.FLAGGED);
+            }
+            if (oldKeyword.isUnread().isPresent() && !oldKeyword.isUnread().get()) {
+                builder.add(Keyword.SEEN);
+            }
+            return fromSet(builder.build());
+        }
+
+        public Keywords fromFlags(Flags flags) {
+            return fromSet(Stream.concat(
+                    Stream.of(flags.getUserFlags())
+                        .map(Keyword::new),
+                    Stream.of(flags.getSystemFlags())
+                        .map(Keyword::fromFlag))
+                .collect(Guavate.toImmutableSet()));
+        }
+
+        public Optional<Keywords> fromMapOrOldKeyword(Optional<Map<String, Boolean>> mapKeyword, Optional<OldKeyword> oldKeyword) {
+            Preconditions.checkArgument(!(mapKeyword.isPresent() && oldKeyword.isPresent()), "Does not support keyword and is* at the same time");
+
+            Keywords keywords = mapKeyword.map(this::fromMap)
+                .orElse(oldKeyword.map(this::fromOldKeyword)
+                    .orElse(null));
+            return Optional.ofNullable(keywords);
```

Comment:
In update case, we had isFlagIdentity which should be check in case of we do not change flags --> that's why I would like to serialization keywords as Optional